pkg/v1/webhooks/webhooks.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2022 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+// Package webhooks provides functions to manage webhook TLS certificates
+package webhooks
+
+import (
+	"context"
+	"strings"
+	"time"
+
+	"github.com/go-logr/logr"
+	corev1 "k8s.io/api/core/v1"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/client-go/rest"
+	"knative.dev/pkg/webhook/certificates/resources"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+)
+
+type WebhookTLS struct {
+	Ctx           context.Context
+	K8sConfig     *rest.Config
+	CertPath      string
+	KeyPath       string
+	Name          string
+	ServiceName   string
+	LabelSelector string
+	Logger        logr.Logger
+	secret        *corev1.Secret
+	Namespace     string
+	RotationTime  time.Duration
+}
+
+const oneWeek = time.Hour * 24 * 7
+
+func (w *WebhookTLS) UpdateOrCreate() error {
+	if w.RotationTime > oneWeek {
+		w.Logger.Info("rotation time will be set to maximum allowed value of one Week")
+		w.RotationTime = oneWeek
+	}
+	gracePeriod := oneWeek - w.RotationTime // Time left to end of one week when certificates need to be rotated.
+	clusterClient, err := client.New(w.K8sConfig, client.Options{})
+	if err != nil {
+		return err
+	}
+	currentSecret := &corev1.Secret{}
+	err = clusterClient.Get(w.Ctx, client.ObjectKey{
+		Namespace: w.Namespace,
+		Name:      w.Name}, currentSecret)
+	if err == nil {
+		w.secret = currentSecret
+	} else if !apierrors.IsNotFound(err) { // secret not found = "Create" case.
+		return err
+	}
+
+	err = ValidateTLSSecret(w.secret, gracePeriod)
+	if err != nil {
```

Comment:
question: so any error results is new cert?

Reply:
yes.  We could add more specific tests. Also perhaps I will make it return a boolean instead and we can log if an error.

Reply:
expanded ValidateTLSSecret to return a boolean and an error. currently any error will result on new certificate, but we can augment if required based on the retruned logic and error.